go_util/util.go

```diff
@@ -31,3 +36,167 @@ func AppendPendingMetadata(testName string, link shared.MetadataLink, metadata *
 		metadata.Links = append(metadata.Links, link)
 	}
 }
+
+// WriteMetadataLink adds a metadataLink to a wpt-metadata META.yml
+// entry and writes it to disk.
+//
+// Directories and META.yml files are created as needed, or an existing one may
+// be updated.
+func WriteMetadataLink(testName string, link shared.MetadataLink) {
+	dirpath := path.Dir(testName)
+	os.MkdirAll(dirpath, os.ModePerm)
+
+	// Open the META.yml file that we would write to, if it exists. In the case
+	// where it does exist we don't want to overwrite the file, so open in
+	// read-write-or-create mode.
+	filepath := dirpath + "/META.yml"
+	f, err := os.OpenFile(filepath, os.O_RDWR|os.O_CREATE, 0644)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer f.Close()
+
+	fi, err := f.Stat()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	if fi.Size() == 0 {
+		// This is an entirely new META.yml file, so create a new
+		// shared.Metadata representation for our expectation and write it out.
+		fmt.Println("CREATED-META-YML: for test", testName, "with url", link.URL)
+		metadata := shared.Metadata{
+			Links: []shared.MetadataLink{link},
+		}
+		writeToFile(metadata, f)
+		return
+	}
+
+	data, err := ioutil.ReadAll(f)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	var metadata shared.Metadata
+	err = yaml.Unmarshal(data, &metadata)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	appendMetadataLink(testName, link, &metadata)
+	writeToFile(metadata, f)
+}
+
+// deleteMetadata deletes testname from metadata and return the new metadata and deleted metadataLink entries.
+func deleteMetadata(testname string, metadata shared.Metadata) (shared.Metadata, shared.MetadataLinks) {
```

Comment:
Is the `testname` here the same kind of value as the `testPath` argument to `DeleteTestFromMetadata`? If so it would be more clear to name this `testPath` too, and if not a comment explaining what the string is in both cases would be great.

(Maybe it's the distinction between test path and file path, I'm not sure without digging deep.)

Reply:
`testname` is the `path.Base(testpath)` so yes they are different. I added some comments